### PR TITLE
operation_logs: capture operations

### PR DESCRIPTION
**Pull request by BadrAly**
* Captures operation after record creation of the following resources:
    * documents
    * holdings
    * items
* Adds units testing.

Co-Authored-by: Aly Badr <aly.badr@rero.ch>

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

File: rero_ils/modules/operation_logs/listener.py [hunk 1]
```diff
@@ -22,37 +22,61 @@
...
+    build_operation_log_record(
+        record=record, operation=OperationLogOperation.CREATE)
+
+
+def operation_log_record_update(sender, record=None, *args, **kwargs):
+    """Create an operation log entry after record update.
+
+    Checks if the record is configured to keep logs of its update.
+    If enabled, a new operation log entry will be added. Method is called after
+    record update by connecting to signals'after_record_update'.
```
Comment (iGormilhit) [inline rero_ils/modules/operation_logs/listener.py:48]:
```suggestion
    record update by connecting to the 'after_record_update' signals.
```

File: tests/api/operation_logs/test_operation_logs_permissions.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2021 RERO
```
Comment (iGormilhit) [inline tests/api/operation_logs/test_operation_logs_permissions.py:4]: :+1: :slightly_smiling_face:

File: tests/ui/operation_logs/test_operation_logs_jsonresolver.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline tests/ui/operation_logs/test_operation_logs_jsonresolver.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: tests/ui/operation_logs/test_operation_logs_mapping.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline tests/ui/operation_logs/test_operation_logs_mapping.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

**Review by iGormilhit — CHANGES_REQUESTED** [review]
Commit message proposition:

```
operation_logs: capture operations

* Captures operation after record creation of the following resources:
    * documents
    * holdings
    * items
* Adds units testing.
```

**Review by iGormilhit — APPROVED** [review]

**Review by lauren-d — APPROVED** [review]

**Review by benerken — APPROVED** [review]

File: rero_ils/modules/ext.py [hunk 5]
```diff
@@ -152,7 +153,8 @@ def register_signals(self, app):
         after_record_insert.connect(create_subscription_patron_transaction)
         after_record_update.connect(create_subscription_patron_transaction)
         
-        after_record_insert.connect(create_operation_log_record)
+        after_record_insert.connect(operation_log_record_create)
+        after_record_update.connect(operation_log_record_update)
```
Comment (zannkukai) [inline rero_ils/modules/ext.py:157]: create an issue/tasks to implement 'delete' operation with the signal 'before_record_delele' ?

File: rero_ils/modules/operation_logs/listener.py [hunk 6]
```diff
@@ -22,37 +22,61 @@
...
+
+    :param record: the record being created or updated.
+    """
     if record.get('$schema'):
         resource_name = extracted_data_from_ref(
             record.get('$schema'), data='resource')
-        if resource_name in current_app.config.get(
-            'RERO_ILS_ENABLE_OPERATION_LOG'):
+        if resource_name in [resource for resource in current_app.config.get(
+            'RERO_ILS_ENABLE_OPERATION_LOG')]:
```
Comment (zannkukai) [inline rero_ils/modules/operation_logs/listener.py:65]: `if resource_name in current_app.config.get('RERO_ILS_ENABLE_OPERATION_LOG', {}).keys()` seems better

File: rero_ils/modules/operation_logs/permissions.py [hunk 7]
```diff
@@ -32,11 +32,11 @@ def list(cls, user, record=None):
...
         :return: "True" if action can be done.
         """
-        # all users may list operation_log records.
-        # TODO: check with group PO here.
+        # all users (lib, sys_lib) can read operation_log records.
         if not current_patron:
             return False
-        return True
+        if current_patron.is_librarian:
+            return True
```
Comment (zannkukai) [inline rero_ils/modules/operation_logs/permissions.py:39]: `return current_patron.is_librarian`.
With this implementation, a user with only "patron" role, this function return `None` instead False

File: rero_ils/modules/operation_logs/permissions.py [hunk 8]
```diff
@@ -46,11 +46,11 @@ def read(cls, user, record):
...
         :return: "True" if action can be done.
         """
-        # all users may read operation_log records.
-        # TODO: check with group PO here.
+        # all users (lib, sys_lib) can read operation_log records.
         if not current_patron:
             return False
-        return True
+        if current_patron.is_librarian:
+            return True
```
Comment (zannkukai) [inline rero_ils/modules/operation_logs/permissions.py:53]: same here

File: tests/api/operation_logs/test_operation_logs_permissions.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+from flask import url_for
+from invenio_accounts.testutils import login_user_via_session
+from utils import get_json
+
+from rero_ils.modules.operation_logs.api import OperationLog
+
+
+def test_operation_logs_permissions_api(
```
Comment (zannkukai) [inline tests/api/operation_logs/test_operation_logs_permissions.py:25]: missing the test functions to test `OperationLogPermission` class without using the API

**Review by zannkukai — CHANGES_REQUESTED** [review]

File: rero_ils/modules/ext.py (hunk 5 above)
Comment (BadrAly) [reply rero_ils/modules/ext.py:157]: The current Virtua `039` implementation captures only add/update, I will pass the info to @pronguen to decide issue/ticket/task or US.

**Review by zannkukai — APPROVED** [review]

File: rero_ils/modules/ext.py (hunk 5 above)
Comment (pronguen) [reply rero_ils/modules/ext.py:157]: To be done later. It requires a deeper analysis and also to keep the records with a status `deleted`.

**Pull request merged**